File: connector.cpp
#include "connector.hh"

#include <iostream>
#include <ctime>
#include <sstream>
#include <cstdint>
#include <vector>
#include <cstring>

#ifdef WIN32

#include <winsock2.h>
#include <ws2tcpip.h>
#pragma comment(lib, "Ws2_32.lib")
#pragma comment(lib, "Mswsock.lib")
#pragma comment(lib, "AdvApi32.lib")

#else

#include <errno.h>
#include <netdb.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>

#endif

using std::vector;

using cpprofiler::NodeStatus;

namespace Profiling {

Node::Node(int sid, int pid, int alt, int kids, NodeStatus status, Connector& c)
    : _c(c), sid_(sid), pid_(pid), alt_(alt), kids_(kids), status_(status) {}

void Node::send() { this->_c.sendNode(*this); }

Connector::Connector(unsigned int port, unsigned int wid) : port(port), _connected(false) {}

// From http://beej.us/guide/bgnet/output/html/multipage/advanced.html#sendall
static int sendall(int s, const char* buf, int* len) {
  int total = 0;         // how many bytes we've sent
  int bytesleft = *len;  // how many we have left to send
  int n;

  while (total < *len) {
    n = send(s, buf + total, bytesleft, 0);
    if (n == -1) {
      break;
    }
    total += n;
    bytesleft -= n;
  }

  *len = total;  // return number actually sent here

  return n == -1 ? -1 : 0;  // return -1 on failure, 0 on success
}

void Connector::sendOverSocket() {
  if (!_connected) return;

  vector<char> buf = marshalling.serialize();

  sendRawMsg(buf);
}

void Connector::sendRawMsg(const vector<char>& buf) {
  uint32_t bufSize = buf.size();
  int bufSizeLen = sizeof(uint32_t);
  sendall(sockfd, reinterpret_cast<char*>(&bufSize), &bufSizeLen);
  int bufSizeInt = bufSize;
  sendall(sockfd, reinterpret_cast<const char*>(buf.data()), &bufSizeInt);
}


Node Connector::createNode(int sid, int pid, int alt, int kids,
                           NodeStatus status) {
  return Node(sid, pid, alt, kids, status, *this);
}

void Connector::sendNode(const Profiling::Node& node) {

  if (!_connected) return;

  auto& msg = marshalling.makeNode(node.sid(), node.pid(), node.alt(), node.kids(), node.status());

  if (node.label().valid()) msg.set_label(node.label().value());

  if (node.nogood().valid()) msg.set_nogood(node.nogood().value());

  if (node.info().valid()) msg.set_info(node.info().value());

  if (node.solution().valid()) msg.set_solution(node.solution().value());

  if (node.tid().valid()) msg.set_tid(node.tid().value());

  if (node.rid().valid()) msg.set_rid(node.rid().value());

  sendOverSocket();
}

void Connector::restart(const std::string& file_path, int rid,
                        const std::string& variableList, int ex_id) {
  /// extract fzn file name
  std::string label(file_path);
  {
    int pos = label.find_last_of('/');
    if (pos > 0) {
      label = label.substr(pos + 1, label.length() - pos - 1);
    }
  }

  std::string info{""};
  {
    std::stringstream ss;
    ss << "{";
    if (ex_id != -1) {
      ss << "\"execution_id\": " << ex_id;
    }
    if (variableList.size() > 0) {
      if (ex_id != -1) ss << ", ";
      ss << "\"variable_list\": \"" << variableList << "\"";
    }
    ss << "}";
    info = ss.str();
  }

  marshalling.makeStart(rid, label, info);
  sendOverSocket();
}

void Connector::connect() {
  struct addrinfo hints, *servinfo, *p;
  int rv;

#ifdef WIN32
  // Initialise Winsock.
  WSADATA wsaData;
  int startupResult = WSAStartup(MAKEWORD(2, 2), &wsaData);
  if (startupResult != 0) {
    printf("WSAStartup failed with error: %d\n", startupResult);
  }
#endif

  memset(&hints, 0, sizeof hints);
  hints.ai_family = AF_UNSPEC;
  hints.ai_socktype = SOCK_STREAM;

  if ((rv = getaddrinfo("localhost", std::to_string(port).c_str(), &hints,
                        &servinfo)) != 0) {
    std::cerr << "getaddrinfo: " << gai_strerror(rv) << "\n";
    goto giveup;
  }

  // loop through all the results and connect to the first we can
  for (p = servinfo; p != NULL; p = p->ai_next) {
    if ((sockfd = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1) {
      // errno is set here, but we don't examine it.
      continue;
    }

    if (::connect(sockfd, p->ai_addr, p->ai_addrlen) == -1) {
#ifdef WIN32
      closesocket(sockfd);
#else
      close(sockfd);
#endif
      // errno is set here, but we don't examine it.
      continue;
    }

    break;
  }

  // Connection failed; give up.
  if (p == NULL) {
    goto giveup;
  }

  freeaddrinfo(servinfo);  // all done with this structure

  _connected = true;

  return;
giveup:
  _connected = false;
  return;
}

void Connector::done() {
  marshalling.makeDone();
  sendOverSocket();
}

void Connector::disconnect() {}

std::ostream& operator<<(std::ostream& os, Node& n) {
  return os << "print node here";
}


}

File: Connector.java
import java.io.File;
import java.io.IOException;
import java.net.Socket;
import java.util.List;

public class Connector {

    public static class Option<T> {
        private T value;
        private boolean present = false;

        public boolean isValid() {
            return present;
        }

        public void set(T value) {
            this.value = value;
            present = true;
        }

        public void unset() {
            present = false;
        }

        public T getValue() {
            if (!present) throw new IllegalStateException("Value not present");
            return value;
        }
    }

    public static class Node {
        private final Connector connector;
        private final int node;
        private final int parent;
        private final int restartCount;
        private final int alt;
        private final int kids;
        private final Message.NodeStatus status;
        private final Option<String> label = new Option<>();
        private final Option<String> nogood = new Option<>();
        private final Option<String> info = new Option<>();

        public Node(int node, int parent, int restartCount, int alt, int kids, Message.NodeStatus status, Connector connector) {
            this.connector = connector;
            this.node = node;
            this.parent = parent;
            this.restartCount = restartCount;
            this.alt = alt;
            this.kids = kids;
            this.status = status;
        }

        public Option<String> getLabel() {
            return label;
        }

        public Node setLabel(String label) {
            this.label.set(label);
            return this;
        }

        public Option<String> getNogood() {
            return nogood;
        }

        public Node setNogood(String nogood) {
            this.nogood.set(nogood);
            return this;
        }

        public Option<String> getInfo() {
            return info;
        }

        public Node setInfo(String info) {
            this.info.set(info);
            return this;
        }

        public int getAlt() {
            return alt;
        }

        public int getKids() {
            return kids;
        }

        public Message.NodeStatus getStatus() {
            return status;
        }

        public void send() throws IOException {
            System.out.printf("Sending node %d\n", node);
            Connector.sendNode(connector, this);
        }
    }

    private final Message.MessageMarshalling marshalling;
    private final int port;
    private Socket socket;
    private boolean connected;

    public Connector(int port) {
        this.port = port;
        this.connected = false;
        this.marshalling = new Message.MessageMarshalling();
    }

    public boolean isConnected() {
        return connected;
    }

    public void connect() throws IOException {
        socket = new Socket("localhost", port);
        connected = true;
    }

    public void sendRawMsg(byte[] buf) throws IOException {
        // Each message starts with a four-byte integer encoding the size of the remainder of the message in bytes
        int bufSize = buf.length;
        byte[] sizeBuffer = new byte[4 + bufSize];
        for (int i = 0; i < 4; i++) {
            sizeBuffer[i] = (byte) (bufSize >>> (i * 8));
        }
        System.arraycopy(buf, 0, sizeBuffer, 4, bufSize);
        socket.getOutputStream().write(sizeBuffer);
        socket.getOutputStream().flush();
    }

    public void start(String filePath, int executionId, boolean hasRestarts) throws IOException {
        String baseName = new File(filePath).getName();

        String info = String.format(
                "{\"has_restarts\": %b, \"name\": \"%s\", \"execution_id\": %d}",
                hasRestarts, baseName, executionId
        );
        marshalling.makeStart(info);
        sendOverSocket();
    }

    public void restart(int restartId) throws IOException {
        String info = String.format("{\"restart_id\": %d}", restartId);
        marshalling.makeRestart(info);
        sendOverSocket();
    }

    public void done() throws IOException {
        marshalling.makeDone();
        sendOverSocket();
    }

    public void disconnect() throws IOException {
        if (socket != null && !socket.isClosed()) {
            socket.close();
        }
    }

    private void sendOverSocket() throws IOException {
        if (!connected) return;

        List<Byte> buf = marshalling.serialize();
        byte[] byteArray = new byte[buf.size()];
        for (int i = 0; i < buf.size(); i++) {
            byteArray[i] = buf.get(i);
        }

        sendRawMsg(byteArray);
    }

    public void sendNode(Node node) throws IOException {
        if (!connected) return;

        Message msg = marshalling.makeNode(node.node, node.parent, node.restartCount,
                node.getAlt(), node.getKids(), node.getStatus());

        if (node.getLabel().isValid()) msg.setLabel(node.getLabel().getValue());
        if (node.getNogood().isValid()) msg.setNogood(node.getNogood().getValue());
        if (node.getInfo().isValid()) msg.setInfo(node.getInfo().getValue());
        sendOverSocket();
    }

    public Node createNode(int node, int parent, int restartCount, int alt, int kids, Message.NodeStatus status) {
        return new Node(node, parent, restartCount, alt, kids, status, this);
    }

    public static void sendNode(Connector connector, Node node) throws IOException {
        connector.sendNode(node);
    }
}
